File: src/agent/agent-sdk/python/bkci_agent_sdk/worker_upgrade.py
"""Validated, atomic worker-agent.jar initialization and upgrade."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
import os
from pathlib import Path
import shutil
from typing import Callable, Mapping

from .download import WORK_AGENT_FILE, download_worker_jar, file_md5
from .worker import detect_worker_version


@dataclass(slots=True)
class WorkerJarState:
    version: str
    md5: str


@dataclass(slots=True)
class WorkerJarUpgradeResult(WorkerJarState):
    changed: bool
    not_modified: bool


class DefaultWorkerJarManager:
    def __init__(
        self,
        *,
        gateway: str,
        auth_headers: Mapping[str, str],
        work_dir: str | Path,
        jdk17_path: str | Path,
        worker_jar_path: str | Path | None = None,
        upgrade_dir: str | Path | None = None,
        timeout_seconds: float = 300,
        log_fn: Callable[[str], None] | None = None,
    ) -> None:
        self.gateway = gateway
        self.auth_headers = dict(auth_headers)
        self.work_dir = Path(work_dir).resolve()
        self.jdk17_path = jdk17_path
        self.worker_jar_path = Path(
            worker_jar_path or self.work_dir / WORK_AGENT_FILE
        ).resolve()
        self.upgrade_dir = Path(upgrade_dir or self.work_dir / "upgrade").resolve()
        self.timeout_seconds = timeout_seconds
        self.log = log_fn or (lambda message: None)
        self._version = ""
        self._active_upgrade: asyncio.Task[WorkerJarUpgradeResult] | None = None
        if self.upgrade_dir / WORK_AGENT_FILE == self.worker_jar_path:
            raise ValueError("upgrade_dir must not contain the active worker-agent.jar")

    def get_version(self) -> str:
        return self._version

    def get_worker_jar_path(self) -> str:
        return str(self.worker_jar_path)

    async def initialize(self) -> WorkerJarState:
        version = await self._detect_version(self.worker_jar_path, self.worker_jar_path.parent)
        if version:
            self._version = version
            md5 = await file_md5(self.worker_jar_path)
            self.log(f"worker initialized: version={version}, md5={md5}")
            return WorkerJarState(version, md5)
        self.log("active worker is missing or invalid, downloading the default worker-agent.jar")
        result = await self.upgrade()
        return WorkerJarState(result.version, result.md5)

    async def upgrade(self) -> WorkerJarUpgradeResult:
        if self._active_upgrade is None or self._active_upgrade.done():
            self._active_upgrade = asyncio.create_task(
                self._do_upgrade(), name="bkci-worker-upgrade"
            )
        return await asyncio.shield(self._active_upgrade)

    async def _do_upgrade(self) -> WorkerJarUpgradeResult:
        self.upgrade_dir.mkdir(parents=True, exist_ok=True)
        download = await download_worker_jar(
            self.gateway,
            self.auth_headers,
            self.upgrade_dir,
            self.timeout_seconds,
        )
        candidate = self.upgrade_dir / WORK_AGENT_FILE
        version = await self._detect_version(candidate, self.upgrade_dir)
        if not version:
            raise RuntimeError(f"downloaded worker version is invalid: {candidate}")
        current_md5 = await file_md5(self.worker_jar_path)
        changed = current_md5 != download.md5
        if changed:
            await _atomic_copy_file(candidate, self.worker_jar_path, download.md5)
            self.log(
                f"worker upgraded: version={version}, oldMd5={current_md5 or '<missing>'}, "
                f"newMd5={download.md5}"
            )
        else:
            self.log(f"worker already up-to-date: version={version}, md5={download.md5}")
        self._version = version
        return WorkerJarUpgradeResult(
            version=version,
            md5=download.md5,
            changed=changed,
            not_modified=download.not_modified,
        )

    async def _detect_version(self, worker_jar_path: Path, work_dir: Path) -> str:
        return await detect_worker_version(
            jdk17_path=self.jdk17_path,
            worker_jar_path=worker_jar_path,
            work_dir=work_dir,
            log_fn=self.log,
        )


async def _atomic_copy_file(source: Path, target: Path, expected_md5: str) -> None:
    target.parent.mkdir(parents=True, exist_ok=True)
    temporary = target.with_name(f"{target.name}.upgrade.{os.getpid()}.{id(asyncio.current_task())}")
    try:
        await asyncio.to_thread(shutil.copyfile, source, temporary)
        copied_md5 = await file_md5(temporary)
        if copied_md5 != expected_md5:
            raise RuntimeError(
                f"copied worker md5 not match: expected={expected_md5}, actual={copied_md5}"
            )
        await asyncio.to_thread(os.replace, temporary, target)
    finally:
        temporary.unlink(missing_ok=True)

File: src/agent/agent-sdk/nodejs/src/workerUpgrade.ts
/*
 * BK-CI Agent SDK - 默认 worker-agent.jar 管理与升级
 *
 * 对齐 Go agent 的 pkg/upgrade/downloadUpgradeWorker + DoUpgradeOperation：
 *   1. 下载到 <workDir>/upgrade/worker-agent.jar，不直接覆盖正在使用的 jar；
 *   2. 使用 JDK17 运行 AgentVersionKt，候选版本合法后才允许替换；
 *   3. 比较 MD5，内容未变化时跳过替换；
 *   4. 通过同目录临时文件 + rename 更新工作目录中的 worker-agent.jar。
 *
 * 是否有构建运行、升级期间是否停止接单以及 finishUpgrade 上报由 AgentHandler
 * 编排；这样该管理器也可以脱离 AgentLoop 单独使用。
 */

import * as fs from 'fs';
import * as path from 'path';
import { WorkAgentFile, downloadWorkerJar, fileMd5 } from './download';
import { detectWorkerVersion } from './worker';

export interface DefaultWorkerJarManagerOptions {
  /** 后台网关地址。 */
  gateway: string;
  /** 下载 worker-agent.jar 使用的 Agent 鉴权头。 */
  authHeaders: Record<string, string>;
  /** Agent 工作目录。 */
  workDir: string;
  /** JDK17 的 java 可执行文件路径或 JDK 目录。 */
  jdk17Path: string;
  /** 正式 worker-agent.jar 路径；默认 <workDir>/worker-agent.jar。 */
  workerJarPath?: string;
  /** 候选升级文件目录；默认 <workDir>/upgrade。 */
  upgradeDir?: string;
  /** 下载超时（毫秒），默认使用 downloadWorkerJar 的 300000。 */
  timeoutMs?: number;
  /** 可选日志回调，默认走 console。 */
  logFn?: (msg: string) => void;
}

export interface WorkerJarState {
  version: string;
  md5: string;
}

export interface WorkerJarUpgradeResult extends WorkerJarState {
  /** 正式 worker-agent.jar 是否被替换。 */
  changed: boolean;
  /** 下载接口是否返回 304。 */
  notModified: boolean;
}

/**
 * 官方 worker-agent.jar 的默认管理器。
 *
 * initialize() 在启动前获取当前版本；当前 jar 缺失或无合法版本时，会尝试从后台
 * 下载并安装默认 jar。upgrade() 用于处理后台下发的 worker 升级指令。
 */
export class DefaultWorkerJarManager {
  private readonly opts: DefaultWorkerJarManagerOptions;
  private readonly workerJarPath: string;
  private readonly upgradeDir: string;
  private readonly log: (msg: string) => void;
  private version = '';
  private activeUpgrade?: Promise<WorkerJarUpgradeResult>;

  constructor(opts: DefaultWorkerJarManagerOptions) {
    this.opts = opts;
    this.workerJarPath = path.resolve(
      opts.workerJarPath ?? path.join(opts.workDir, WorkAgentFile)
    );
    this.upgradeDir = path.resolve(opts.upgradeDir ?? path.join(opts.workDir, 'upgrade'));
    this.log = opts.logFn ?? ((m) => console.info('[agent-sdk][worker-upgrade]', m));

    const upgradeJarPath = path.join(this.upgradeDir, WorkAgentFile);
    if (upgradeJarPath === this.workerJarPath) {
      throw new Error('upgradeDir must not contain the active worker-agent.jar');
    }
  }

  /** 当前已确认合法的 worker 版本；initialize 前为空字符串。 */
  getVersion(): string {
    return this.version;
  }

  /** 正式 worker-agent.jar 的绝对路径。 */
  getWorkerJarPath(): string {
    return this.workerJarPath;
  }

  /**
   * 启动初始化：优先使用当前 jar；缺失或版本非法时从后台下载默认 jar 自愈。
   * 初始化失败会抛异常，避免 Agent 用空版本启动后继续领取构建。
   */
  async initialize(): Promise<WorkerJarState> {
    const version = await this.detectVersion(this.workerJarPath, path.dirname(this.workerJarPath));
    if (version !== '') {
      this.version = version;
      const md5 = await fileMd5(this.workerJarPath);
      this.log(`worker initialized: version=${version}, md5=${md5}`);
      return { version, md5 };
    }

    this.log('active worker is missing or invalid, downloading the default worker-agent.jar');
    const result = await this.upgrade();
    return { version: result.version, md5: result.md5 };
  }

  /**
   * 下载、校验并按需替换默认 worker-agent.jar。
   * 同一个管理器上的并发调用会复用同一次升级，避免重复下载和相互覆盖。
   */
  upgrade(): Promise<WorkerJarUpgradeResult> {
    if (this.activeUpgrade) {
      return this.activeUpgrade;
    }

    const task = this.doUpgrade();
    const clearActiveUpgrade = (): void => {
      if (this.activeUpgrade === task) {
        this.activeUpgrade = undefined;
      }
    };
    this.activeUpgrade = task;
    void task.then(clearActiveUpgrade, clearActiveUpgrade);
    return task;
  }

  private async doUpgrade(): Promise<WorkerJarUpgradeResult> {
    await fs.promises.mkdir(this.upgradeDir, { recursive: true });
    const download = await downloadWorkerJar(
      this.opts.gateway,
      this.opts.authHeaders,
      this.upgradeDir,
      this.opts.timeoutMs
    );
    const upgradeJarPath = path.join(this.upgradeDir, WorkAgentFile);
    const version = await this.detectVersion(upgradeJarPath, this.upgradeDir);
    if (version === '') {
      throw new Error(`downloaded worker version is invalid: ${upgradeJarPath}`);
    }

    const currentMd5 = await fileMd5(this.workerJarPath);
    const changed = currentMd5 !== download.md5;
    if (changed) {
      await atomicCopyFile(upgradeJarPath, this.workerJarPath, download.md5);
      this.log(
        `worker upgraded: version=${version}, oldMd5=${currentMd5 || '<missing>'}, newMd5=${download.md5}`
      );
    } else {
      this.log(`worker already up-to-date: version=${version}, md5=${download.md5}`);
    }

    this.version = version;
    return {
      version,
      md5: download.md5,
      changed,
      notModified: download.notModified,
    };
  }

  private detectVersion(workerJarPath: string, workDir: string): Promise<string> {
    return detectWorkerVersion({
      jdk17Path: this.opts.jdk17Path,
      workerJarPath,
      workDir,
      logFn: this.log,
    });
  }
}

/** 将已校验的候选文件复制到目标目录后原子替换正式文件。 */
async function atomicCopyFile(source: string, target: string, expectedMd5: string): Promise<void> {
  await fs.promises.mkdir(path.dirname(target), { recursive: true });
  const tempPath = `${target}.upgrade.${process.pid}.${Date.now()}`;
  try {
    await fs.promises.copyFile(source, tempPath);
    const copiedMd5 = await fileMd5(tempPath);
    if (copiedMd5 !== expectedMd5) {
      throw new Error(`copied worker md5 not match: expected=${expectedMd5}, actual=${copiedMd5}`);
    }
    await fs.promises.rename(tempPath, target);
  } catch (e) {
    await fs.promises.rm(tempPath, { force: true }).catch(() => undefined);
    throw e;
  }
}
